### ISSUE TITLE
MapItemData requires a World and an ItemStack to manipulate properly, but API doesn't allow this cleanly.

### ISSUE THREAD

**Issue by modwizcode** [SpongePowered/SpongeAPI#809]
For some types of `Value<T>` such as the ones associated with the data stored for a map, the `ItemStack`/`DataContainer` used doesn't provide all the needed parameters to actually manipulate the value with a `DataProcessor` during implementation. In this example to get the `MapData` from Minecraft, I must supply both the `ItemStack` and the `World` for the provided map item. This is because maps are stored separately from the world's normal item data, and each world maintains it's own set of map ID's. 

I believe currently vanilla might just use the default data directory from the overworld for all map data files, however that's not guaranteed to be true in the future, and maps are just one example of `Value<T>` types that need more context than the `DataContainer` they operate on in order to be applied.

One solution to this is to add a `ContextualValue<T,C>` with an `get(C context)` method and perhaps always return Optional.absent() in the normal `get()` method. However this solution still feels lacking and I don't have a solution that seems more thought out.

Pinging @SpongePowered/developers for comments.

**Comment by gabizou** [comment]
You do realize that first off, you have more than enough context with [`Value#getKey()`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/data/value/BaseValue.java#L97) and comparing it with any of the keys available in [`Keys`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/data/key/Keys.java#L153). Aside from the fact that you're worrying about a [`DataProcessor`](https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/data/DataProcessor.java) which is an implementation detail, you should be able to not only modify `MapItemData` as necessary, you should be able to retrieve/set the map instance from/to the `ItemStack`. I don't see why `MapData` MUST supply the `ItemStack` since a map instance isn't attached to an `ItemStack` at all, rather, it's attached to a `World` and a renderer of sorts. 

As well, you do realize that a `DataContainer` isn't what you're trying to get `MapItemData` from or offering to in the `DataProcessor`.

I'm trying to see this from your point of view, but I don't see this as an issue. We have multiple `World` dependent data, I don't see how this is any different.

**Comment by modwizcode** [comment]
Unfortunately you haven't solved the need for a world instance **and** an
item stack. The ItemStack tells you which map data file it is. You really
haven't solved this issue at all.

**Comment by modwizcode** [comment]
Unfortunately you haven't solved the need for a world instance **and** an
item stack. The ItemStack tells you which map data file it is in that world's map data file list. Currently, keeping complete compatibilty with vanilla I could have a `Value<String>` called `ATTACHED_MAP` that can be retrieved on filled map `ItemStack`s and a method on `World` that does `MapItemData getMapData(String mapID)` While not a bad solution it's not what I'd prefer, this should be something that can be manipulated on the `ItemStack` and the hanging frame entity (really a wrapper for get `ItemStack` and do the same operation) directly.

**Comment by modwizcode** [comment]
Never mind this, I think I've got a few thoughts on a solution that can fit within the constructs already available. It's not as nice as I would like, more interfaces are involved, but it should work out correctly.

**Comment by gabizou** [comment]
@modwizcode If a `World` instance is required, can easily attach information to the `NBTTagCompound` for the world's `UUID`, of which the `DataProcessor` accesses, just in the event that we absolutely need to have the world reference to retrieve. Alternatively, can easily hook into the `ItemMap` implementation to generate the appropriate `MapItemData` as necessary. Of course, it's a hack of it's own system.

**Comment by gabizou** [comment]
I'm just going to clarify here: `MapItemData` is supposed to contain an instance of `MapData` or whatever you're going to call it in the API. That's all. Nothing more, nothing less.

**Comment by modwizcode** [comment]
The world I need isn't static though it's whatever the world the ItemStack
is currently in. I can hack the system, but that isn't as compatible with
vanilla. I'm struggling with a good way to solve that besides just
attaching the map data to world and always passing it to a method with an
ItemStack

**Comment by gabizou** [comment]
I'm pretty certain that once the `ItemStack` has been filled with a map, it's static to that world's `MapData`. I'm also fairly certain that you DON'T want to be using the current minecraft implementation of maps to be the "guiding light" so to speak of how Maps should be API'ed. Besides the fact that we already know that all `World`s actually use the main world storage (we're not like bukkit where we store maps as whole different map folders), you already can effectively retrieve the actual map from the `MapStorage` regardless of the dimension id the map was created in. So, sorry, there even then isn't a reason to have a World instance in the `MapItemData`.

**Comment by modwizcode** [comment]
I'm talking about having map methods attached to a world. I was just
interested in your thoughts on the best way to do this. I also was confused
because Minecraft doesn't actually use the dimension stored in the map,
except to compare before updating the map, which isn't the same as fixing
the map to a dimension since the maps stores it's dimension data in the map
data. The only field the itemstack holds is the is scaling property and the
damage value which indicates the specific map in the dimension to load.

In short I'll be rewriting the internals of maps after talking this through
on IRC. If this issue isn't always closed it can be closed now (I can't do
that from email).
